salt/proxy/nxos.py

```diff
@@ -120,462 +195,352 @@ def __virtual__():
-    if encrypted is False:
-        if crypt_salt is None:
-            # NXOS does not like non alphanumeric characters.  Using the random module from pycrypto
-            # can lead to having non alphanumeric characters in the salt for the hashed password.
-            crypt_salt = secure_password(8, use_random=False)
-        hashed_pass = gen_hash(
-            crypt_salt=crypt_salt, password=password, algorithm=algorithm
-        )
-    else:
-        hashed_pass = password
-    password_line = "username {0} password 5 {1}".format(username, hashed_pass)
-    if role is not None:
-        password_line += " role {0}".format(role)
-    try:
-        sendline("config terminal")
-        ret = sendline(password_line)
-        sendline("end")
-        sendline("copy running-config startup-config")
-        return "\n".join([password_line, ret])
-    except TerminalException as e:
-        log.error(e)
-        return "Failed to set password"
-
-
-def remove_user(username):
-    """
-    Remove user from switch
+    method:
+        ``cli_show_ascii``: Return raw test or unstructured output.
+        ``cli_show``: Return structured output.
+        ``cli_conf``: Send configuration commands to the device.
+        Defaults to ``cli_show_ascii``.
 
-    .. code-block:: bash
+        NOTES for SSH proxy minon:
+          ``method`` is ignored for SSH proxy minion.
+          Only show commands are supported and data is returned unstructured.
+          This function is preserved for backwards compatibilty.
 
-        salt '*' nxos.cmd remove_user username=daniel
-    """
-    try:
-        sendline("config terminal")
-        user_line = "no username {0}".format(username)
-        ret = sendline(user_line)
-        sendline("end")
-        sendline("copy running-config startup-config")
-        return "\n".join([user_line, ret])
-    except TerminalException as e:
-        log.error(e)
-        return "Failed to set password"
-
-
-def set_role(username, role):
-    """
-    Assign role to username
+    .. code-block: bash
 
-    .. code-block:: bash
-
-        salt '*' nxos.cmd set_role username=daniel role=vdc-admin
+        salt '*' nxos.cmd sendline 'show run | include "^username admin password"'
     """
     try:
-        sendline("config terminal")
-        role_line = "username {0} role {1}".format(username, role)
-        ret = sendline(role_line)
-        sendline("end")
-        sendline("copy running-config startup-config")
-        return "\n".join([role_line, ret])
-    except TerminalException as e:
+        if CONNECTION == "ssh":
+            result = _sendline_ssh(command, **kwargs)
+        elif CONNECTION == "nxapi":
+            result = _nxapi_request(command, method, **kwargs)
+    except (TerminalException, NxosCliError) as e:
         log.error(e)
-        return "Failed to set password"
+        raise
+    return result
 
 
-def unset_role(username, role):
+def proxy_config(commands, save_config=None):
```

Comment:
Why are we calling this proxy_config instead of config?

To be consistent, we would also call the proxy.sendline proxy_sendline, etc.

Reply:
Ok, I know where some of the confusion is coming from.  The only functions that are really callable are the functions defined in the execution module `salt/modules/nxos.py`.  The functions in `salt/proxy/nxos.py` are called by functions in `salt/modules/nxos.py` when a proxy is needed.

That means that `proxy_config` is not directly callable.

```shell
(salt_py375) ➜  salt_py375 salt -c ./etc/salt 'n9k-proxy' nxos.p
n9k-proxy:
    ----------
    nxos.p:
        'nxos.p' is not available.
    nxos.ping:

            Ping the device on the other end of the connection.

            .. code-block: bash

                salt '*' nxos.ping

ERROR: Minions returned with non-zero exit code
(salt_py375) ➜  salt_py375 salt -c ./etc/salt 'n9k-proxy' nxos.proxy_config
n9k-proxy:
    'nxos.proxy_config' is not available.
ERROR: Minions returned with non-zero exit code
(salt_py375) ➜  salt_py375
```

Before my refactoring most of the functions were in the proxy source file and where directly callable but to streamline the workflows for all minion types I refactored the execution flow to always go through the execution module.  In order to `expose` a function in the proxy source file we need a corresponding function in the execution module source file.

From my perspective we need to do the following.

1. Remove function level documentation in the proxy source file and/or add a note to indicate that functions in the proxy file are documented in the module source file.
2. Add a function in the module source file for the `shutdown` function and return a message when `_shutdown_nxapi` is called to indicate that shutting down the nxapi minion connection is not possible.

Reply:
**Update:** I decided to remove the `shutdown` function because it does not provide any value.

Reply:
Never mind.  Turns out that a `shutdown` function is required but since it's not adding value I turned them into no-ops.

```
The Salt ProxyMinion is shutdown.
Proxymodule nxos is missing an init() or a shutdown() or both. Check your proxymodule.  Salt-proxy aborted.
```

Reply:
I guess my confusion with the name is because every other function in proxy/nxos has the same name as the equivalent in modules/nxos (like `proxy.sendline`), while this is named `proxy.proxy_config`.

Reply:
I believe the shutdown function is called when shutting down a proxy minion. It seems that explicitly shutting down the connection may be better... You didn't notice any issues when you restarted the ssh proxy minion, I guess?

https://docs.saltstack.com/en/latest/topics/proxyminion/index.html#proxymodules

Reply:
> I believe the shutdown function is called when shutting down a proxy minion. It seems that explicitly shutting down the connection may be better... You didn't notice any issues when you restarted the ssh proxy minion, I guess?
> 
> https://docs.saltstack.com/en/latest/topics/proxyminion/index.html#proxymodules

What exactly do you mean by restart?  I was able to stop and start the minion without any issues.  I also see the following in the guide you sent.

```
shutdown(): Code to cleanly shut down or close a connection to a controlled device goes here. This function must exist, but can contain only the keyword pass if there is no shutdown logic required.
```

I could add the code back that we had for the ssh proxy minion since it's not hurting anything just to be safe but just use `pass` for the nxapi proxy minion.

Reply:
> I guess my confusion with the name is because every other function in proxy/nxos has the same name as the equivalent in modules/nxos (like `proxy.sendline`), while this is named `proxy.proxy_config`.

Just took another look at the execution module functions and I believe we were trying to avoid confusion because of the following.

Functions in `modules/nxos` call `_configure_device` which in turn calls `proxy.proxy_config` if a proxy is needed.  This is also true for the function in `modules/nxos` that is named `config`.  I think we were trying to make it clear that `config` and `proxy_config` are different.  I suppose we could still rename `proxy_config` in `proxy/nxos` to be `config` and there would not be a conflict but it might be more confusing?

Reply:
Let's leave it as is for now :)